File: CDT2026/algo_polyglot/java/SeismoGuardAlgo.java
// SeismoGuardAlgo.java — Java port of the SeismoGuard EEW algorithm.
// Java 17+. Build: javac SeismoGuardAlgo.java && java SeismoGuardAlgo
// Tests embedded in main() — exits non-zero on failure.

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Optional;

public final class SeismoGuardAlgo {

    // ── Canonical constants (see spec/SAMPLE_RECORD.md) ─────────────────
    public static final int    SAMPLE_RATE_HZ  = 100;
    public static final double DT              = 1.0 / SAMPLE_RATE_HZ;
    public static final double ALPHA_STA       = 0.02;
    public static final double ALPHA_LTA       = 0.000333;
    public static final double ALPHA_DC        = 0.001;
    public static final double RATIO_TRIGGER   = 6.0;
    public static final double RATIO_DETRIGGER = 1.5;
    public static final int    MIN_TRIG_COUNT  = 3;
    public static final double SPIKE_LIMIT     = 50.0;
    public static final double LTA_FLOOR       = 1e-9;
    public static final double MIN_TRIG_CF_ABS = 2e-4;
    public static final int    WINDOW_SAMPLES  = 300;
    public static final double HPF_ALPHA       = 0.9901;
    public static final double MW_PD_A         = 0.813;
    public static final double MW_PD_B         = 1.512;
    public static final double MW_PD_C         = 5.130;
    public static final double MW_R_KM_DEFAULT = 10.0;
    public static final int    ALARM_MAX_MS    = 15_000;
    public static final int    LOCKOUT_MAX_MS  =  5_000;

    public enum State { STANDBY, DETECTING, ALARMING, LOCKOUT }

    public static final class SampleRecord {
        public final long   timestampMs;       // u32 range in long
        public final double staLtaRatio;
        public final double cfZ;
        public final double mpdRaw;
        public final int    sampleCount;       // u16 in int

        SampleRecord(long t, double r, double cf, double mpd, int n) {
            this.timestampMs = t & 0xFFFFFFFFL;
            this.staLtaRatio = r;
            this.cfZ         = cf;
            this.mpdRaw      = mpd;
            this.sampleCount = n & 0xFFFF;
        }

        public String toJson() {
            // Precision pinned to match Python/C/Rust:
            //   sta_lta_ratio: 6 decimals
            //   cf_z, mpd_raw: full double precision (Double.toString round-trips)
            return "{\"timestamp_ms\":" + timestampMs
                 + ",\"sta_lta_ratio\":" + String.format(java.util.Locale.ROOT, "%.6f", staLtaRatio)
                 + ",\"cf_z\":" + Double.toString(cfZ)
                 + ",\"mpd_raw\":" + Double.toString(mpdRaw)
                 + ",\"sample_count\":" + sampleCount + "}";
        }
    }

    public static final class Decision {
        public final boolean fireT1, fireT2;
        public final double  mwEst, pdM, tauC;
        public final String  reason;

        Decision(boolean t1, boolean t2, double mw, double pd, double tc, String r) {
            this.fireT1 = t1; this.fireT2 = t2;
            this.mwEst  = mw; this.pdM = pd; this.tauC = tc;
            this.reason = r;
        }
    }

    // ── Instance state ──────────────────────────────────────────────────
    private final double mwT1, mwT2, rKm;
    private State  state;
    private double sta, lta, ltaQuiet;
    private double baselineZ;
    private boolean baselineInit;
    private int    trigCount;
    private double peakRatio;
    private long   alarmStartMs, lockoutStartMs;

    private double hpfAIn, hpfAOut;
    private double vel;
    private double hpfVIn, hpfVOut;
    private double disp;
    private double hpfDIn, hpfDOut;

    private int    winN;
    private double pdMax, tcNum, tcDen;
    private boolean windowOpen;

    private long sampleCount;
    private long startMs;

    public SeismoGuardAlgo() { this(3.5, 4.5, MW_R_KM_DEFAULT); }

    public SeismoGuardAlgo(double mwT1, double mwT2, double rKm) {
        this.mwT1 = mwT1; this.mwT2 = mwT2; this.rKm = rKm;
        reset();
    }

    public void reset() {
        state = State.STANDBY;
        sta = LTA_FLOOR; lta = LTA_FLOOR; ltaQuiet = LTA_FLOOR;
        baselineZ = 0; baselineInit = false;
        trigCount = 0; peakRatio = 0;
        alarmStartMs = 0; lockoutStartMs = 0;
        hpfAIn = hpfAOut = 0; vel = 0;
        hpfVIn = hpfVOut = 0; disp = 0;
        hpfDIn = hpfDOut = 0;
        winN = 0; pdMax = 0; tcNum = 0; tcDen = 0;
        windowOpen = false;
        sampleCount = 0; startMs = 0;
    }

    public void seedLta(double noiseFloorCf) {
        double v = Math.max(noiseFloorCf, LTA_FLOOR);
        sta = v; lta = v; ltaQuiet = v;
    }

    public State state() { return state; }

    public ProcessResult process(double ax, double ay, double az, long tMs) {
        if (sampleCount == 0) startMs = tMs;

        // 1. DC tracker
        if (!baselineInit) { baselineZ = az; baselineInit = true; }
        if (state == State.STANDBY || state == State.DETECTING) {
            baselineZ = ALPHA_DC * az + (1.0 - ALPHA_DC) * baselineZ;
        }
        double dz = az - baselineZ;

        // 2. CF + spike clamp (raw retained for anti-tap)
        double ltaRef = (lta > LTA_FLOOR) ? lta : LTA_FLOOR;
        double cfRaw  = dz * dz;
        double cf     = (cfRaw <= SPIKE_LIMIT * ltaRef) ? cfRaw : SPIKE_LIMIT * ltaRef;

        // 3. STA/LTA
        sta = ALPHA_STA * cf + (1.0 - ALPHA_STA) * sta;
        if (state == State.STANDBY) {
            lta = ALPHA_LTA * cf + (1.0 - ALPHA_LTA) * lta;
        }
        double ratio = (lta > LTA_FLOOR) ? (sta / lta) : 0.0;

        // 4. HPF→∫→HPF→∫→HPF
        double aHpf = HPF_ALPHA * (hpfAOut + dz - hpfAIn);
        hpfAIn = dz; hpfAOut = aHpf;
        vel += aHpf * DT;
        double vHpf = HPF_ALPHA * (hpfVOut + vel - hpfVIn);
        hpfVIn = vel; hpfVOut = vHpf;
        disp += vHpf * DT;
        double dHpf = HPF_ALPHA * (hpfDOut + disp - hpfDIn);
        hpfDIn = disp; hpfDOut = dHpf;

        // 5. Pd window
        Decision decision = null;
        if (windowOpen) {
            double ad = Math.abs(dHpf);
            if (ad > pdMax) pdMax = ad;
            tcNum += dHpf * dHpf;
            tcDen += vel * vel;
            winN++;
            if (winN >= WINDOW_SAMPLES) decision = closeWindow();
        }

        // 6. State machine (uses raw cf for anti-tap)
        stepState(ratio, cfRaw, tMs);

        SampleRecord rec = new SampleRecord(
            tMs - startMs, ratio, cf,
            windowOpen ? pdMax : 0.0,
            (int) sampleCount
        );
        sampleCount++;
        return new ProcessResult(rec, Optional.ofNullable(decision));
    }

    private void openWindow() {
        winN = 0; pdMax = 0; tcNum = 0; tcDen = 0; windowOpen = true;
        hpfAIn = hpfAOut = 0; vel = 0;
        hpfVIn = hpfVOut = 0; disp = 0;
        hpfDIn = hpfDOut = 0;
    }

    private Decision closeWindow() {
        double pd = pdMax;
        double tc = (tcDen > 0)
            ? 2.0 * Math.PI * Math.sqrt(tcNum / tcDen)
            : 0.0;
        double mw = estimateMw(pd);
        windowOpen = false;
        if (mw >= mwT2) return new Decision(true, true,  mw, pd, tc, "Mw_ge_T2");
        if (mw >= mwT1) return new Decision(true, false, mw, pd, tc, "Mw_ge_T1");
        return new Decision(false, false, mw, pd, tc, "below_threshold");
    }

    private double estimateMw(double pdM) {
        if (pdM <= 0) return -99.0;
        return MW_PD_A * Math.log10(pdM) + MW_PD_B * Math.log10(rKm) + MW_PD_C;
    }

    private void stepState(double ratio, double cfRaw, long tMs) {
        switch (state) {
            case STANDBY:
                if (ratio >= RATIO_TRIGGER && cfRaw >= MIN_TRIG_CF_ABS) {
                    trigCount++;
                    if (trigCount >= MIN_TRIG_COUNT) {
                        state = State.DETECTING;
                        peakRatio = ratio;
                        openWindow();
                    }
                } else {
                    trigCount = 0;
                }
                break;
            case DETECTING:
                if (ratio > peakRatio) peakRatio = ratio;
                if (ratio < RATIO_DETRIGGER) {
                    state = State.STANDBY;
                    trigCount = 0; peakRatio = 0;
                }
                break;
            case ALARMING:
                if (tMs - alarmStartMs > ALARM_MAX_MS) {
                    state = State.LOCKOUT;
                    lockoutStartMs = tMs;
                }
                break;
            case LOCKOUT:
                if (tMs - lockoutStartMs > LOCKOUT_MAX_MS) {
                    state = State.STANDBY;
                    lta = ltaQuiet; sta = ltaQuiet;
                }
                break;
        }
    }

    public static final class ProcessResult {
        public final SampleRecord record;
        public final Optional<Decision> decision;
        ProcessResult(SampleRecord r, Optional<Decision> d) { this.record = r; this.decision = d; }
    }

    // ── CLI: stdin CSV → stdout JSONL ───────────────────────────────────
    public static void main(String[] args) throws Exception {
        if (args.length > 0 && args[0].equals("--test")) { System.exit(runTests()); return; }
        SeismoGuardAlgo algo = new SeismoGuardAlgo();
        try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in))) {
            String line; long t = 0;
            while ((line = br.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("#")) continue;
                String[] parts = line.split(",");
                if (parts.length < 3) continue;
                try {
                    double ax = Double.parseDouble(parts[0]);
                    double ay = Double.parseDouble(parts[1]);
                    double az = Double.parseDouble(parts[2]);
                    ProcessResult r = algo.process(ax, ay, az, t);
                    System.out.println(r.record.toJson());
                    if (r.decision.isPresent()) {
                        Decision d = r.decision.get();
                        System.out.printf("{\"decision\":\"%s\",\"mw\":%.3f,\"pd_m\":%s,\"tau_c\":%s}%n",
                                          d.reason, d.mwEst, d.pdM, d.tauC);
                    }
                } catch (NumberFormatException ignored) {}
                t += (long) (DT * 1000);
            }
        }
    }

    // ── Inline tests ────────────────────────────────────────────────────
    private static int runTests() {
        int fails = 0;
        fails += check("dc_no_trigger",   testDcNoTrigger());
        fails += check("mw_formula",      testMwFormula());
        fails += check("json_schema",     testJsonSchema());
        fails += check("sample_wrap",     testSampleWrap());
        fails += check("synthetic_fires", testSyntheticFires());
        System.out.println(fails == 0 ? "ALL TESTS OK" : (fails + " TEST FAILURE(S)"));
        return fails;
    }

    private static int check(String name, boolean ok) {
        System.out.println((ok ? "[ok]   " : "[FAIL] ") + name);
        return ok ? 0 : 1;
    }

    private static boolean testDcNoTrigger() {
        SeismoGuardAlgo a = new SeismoGuardAlgo();
        for (int i = 0; i < 500; i++) {
            ProcessResult r = a.process(0, 0, 9.81, i * 10L);
            if (r.decision.isPresent()) return false;
        }
        return a.state() == State.STANDBY;
    }

    private static boolean testMwFormula() {
        SeismoGuardAlgo a = new SeismoGuardAlgo();
        double mw = MW_PD_A * Math.log10(1e-3) + MW_PD_B * Math.log10(10.0) + MW_PD_C;
        return Math.abs(a.estimateMw(1e-3) - mw) < 1e-9;
    }

    private static boolean testJsonSchema() {
        SeismoGuardAlgo a = new SeismoGuardAlgo();
        String s = a.process(0, 0, 9.81, 0).record.toJson();
        return s.contains("timestamp_ms") && s.contains("sta_lta_ratio")
            && s.contains("cf_z") && s.contains("mpd_raw") && s.contains("sample_count");
    }

    private static boolean testSampleWrap() {
        SeismoGuardAlgo a = new SeismoGuardAlgo();
        for (long i = 0; i < 65540; i++) a.process(0, 0, 9.81, i * 10L);
        ProcessResult r = a.process(0, 0, 9.81, 65540L * 10);
        return r.record.sampleCount == 4;
    }

    private static boolean testSyntheticFires() {
        SeismoGuardAlgo a = new SeismoGuardAlgo();
        for (int i = 0; i < 100; i++) a.process(0, 0, 9.81, i * 10L);
        for (int i = 0; i < WINDOW_SAMPLES + 100; i++) {
            double t = i * DT;
            double env = Math.exp(-1.5 * t);
            double v = 2.0 * Math.sin(2 * Math.PI * 3 * t) * env;
            ProcessResult r = a.process(0, 0, 9.81 + v, (100L + i) * 10);
            if (r.decision.isPresent()) return true;
        }
        return false;
    }
}

File: CDT2026/algo_polyglot/python/seismoguard_algo.py
"""
seismoguard_algo.py — reference Python port of the SeismoGuard EEW algorithm.

This is the canonical implementation. Rust/Java/C ports MUST produce
bit-identical (within float-eps) SampleRecord output for the same input.

Run as CLI: pipe `ax,ay,az` triples on stdin, get SampleRecord JSON lines:
    python3 seismoguard_algo.py < accel.csv
"""
from __future__ import annotations

import json
import math
import sys
from dataclasses import dataclass, asdict
from enum import IntEnum
from typing import Iterator, Optional, TextIO

# ─── Canonical constants (see spec/SAMPLE_RECORD.md) ─────────────────────
SAMPLE_RATE_HZ   = 100
DT               = 1.0 / SAMPLE_RATE_HZ
ALPHA_STA        = 0.02
ALPHA_LTA        = 0.000333
ALPHA_DC         = 0.001
RATIO_TRIGGER    = 6.0
RATIO_DETRIGGER  = 1.5
MIN_TRIG_COUNT   = 3
SPIKE_LIMIT      = 50.0
LTA_FLOOR        = 1e-9
MIN_TRIG_CF_ABS  = 2e-4
WINDOW_SAMPLES   = 300
HPF_ALPHA        = 0.9901
MW_PD_A          = 0.813
MW_PD_B          = 1.512
MW_PD_C          = 5.130
MW_R_KM_DEFAULT  = 10.0

ALARM_MAX_MS     = 15000
LOCKOUT_MAX_MS   = 5000

G = 9.80665


class State(IntEnum):
    STANDBY   = 0
    DETECTING = 1
    ALARMING  = 2
    LOCKOUT   = 3


@dataclass
class SampleRecord:
    timestamp_ms: int     # uint32
    sta_lta_ratio: float
    cf_z: float
    mpd_raw: float        # running Pd (m)
    sample_count: int     # uint16, wraps

    def to_json(self) -> str:
        return json.dumps({
            "timestamp_ms":   self.timestamp_ms & 0xFFFFFFFF,
            "sta_lta_ratio":  round(self.sta_lta_ratio, 6),
            "cf_z":           self.cf_z,
            "mpd_raw":        self.mpd_raw,
            "sample_count":   self.sample_count & 0xFFFF,
        }, separators=(",", ":"))


@dataclass
class Decision:
    fire_t1: bool
    fire_t2: bool
    mw_est: float
    pd_m: float
    tau_c: float
    reason: str


class SeismoGuardAlgo:
    """Stateful per-sample detector. Feed accel in m/s², get a SampleRecord.

    Decision(fire_t1/fire_t2, Mw, Pd, τc) returned when the 3 s post-trigger
    window closes (or None on every other sample).
    """

    def __init__(self,
                 mw_t1: float = 3.5,
                 mw_t2: float = 4.5,
                 r_km: float  = MW_R_KM_DEFAULT) -> None:
        self.mw_t1 = mw_t1
        self.mw_t2 = mw_t2
        self.r_km  = r_km
        self.reset()

    def reset(self) -> None:
        self.state = State.STANDBY
        # STA/LTA seeded at LTA_FLOOR — caller may overwrite via seed_lta()
        # after collecting a noise sample on the deployment site.
        self.sta   = LTA_FLOOR
        self.lta   = LTA_FLOOR
        self.lta_quiet = LTA_FLOOR
        self.baseline_z = 0.0
        self.baseline_init = False
        self.trig_count   = 0
        self.peak_ratio   = 0.0
        self.alarm_start_ms   = 0
        self.lockout_start_ms = 0

        # HPF→∫→HPF→∫→HPF state
        self.hpf_a_in = self.hpf_a_out = 0.0
        self.vel       = 0.0
        self.hpf_v_in = self.hpf_v_out = 0.0
        self.disp      = 0.0
        self.hpf_d_in = self.hpf_d_out = 0.0

        # Pd window accumulators
        self.win_n    = 0
        self.pd_max   = 0.0
        self.tc_num   = 0.0
        self.tc_den   = 0.0
        self.window_open = False

        self.sample_count = 0
        self.start_ms     = 0

    # ─── Per-sample processing ──────────────────────────────────────────
    def process(self, ax: float, ay: float, az: float,
                t_ms: int) -> tuple[SampleRecord, Optional[Decision]]:
        if self.sample_count == 0:
            self.start_ms = t_ms

        # 1. DC tracker (Z only — matches S3 firmware)
        if not self.baseline_init:
            self.baseline_z = az
            self.baseline_init = True
        if self.state in (State.STANDBY, State.DETECTING):
            self.baseline_z = ALPHA_DC * az + (1.0 - ALPHA_DC) * self.baseline_z
        dz = az - self.baseline_z

        # 2. CF = dz²; spike clamp applied for STA/LTA filter input only.
        # Raw cf retained for the MIN_TRIG_CF_ABS anti-tap gate (deliberate
        # divergence from S3 firmware: in quiet environments the clamp would
        # squash cf below the anti-tap threshold and block all triggers).
        lta_ref = self.lta if self.lta > LTA_FLOOR else LTA_FLOOR
        cf_raw = dz * dz
        cf = cf_raw if cf_raw <= SPIKE_LIMIT * lta_ref else SPIKE_LIMIT * lta_ref

        # 3. STA/LTA recursive
        self.sta = ALPHA_STA * cf + (1.0 - ALPHA_STA) * self.sta
        if self.state == State.STANDBY:
            self.lta = ALPHA_LTA * cf + (1.0 - ALPHA_LTA) * self.lta
        ratio = (self.sta / self.lta) if self.lta > LTA_FLOOR else 0.0

        # 4. HPF→∫→HPF→∫→HPF (runs every sample so filters stay settled)
        a_hpf = HPF_ALPHA * (self.hpf_a_out + dz - self.hpf_a_in)
        self.hpf_a_in, self.hpf_a_out = dz, a_hpf
        self.vel += a_hpf * DT
        v_hpf = HPF_ALPHA * (self.hpf_v_out + self.vel - self.hpf_v_in)
        self.hpf_v_in, self.hpf_v_out = self.vel, v_hpf
        self.disp += v_hpf * DT
        d_hpf = HPF_ALPHA * (self.hpf_d_out + self.disp - self.hpf_d_in)
        self.hpf_d_in, self.hpf_d_out = self.disp, d_hpf

        # 5. Pd window accumulation
        decision: Optional[Decision] = None
        if self.window_open:
            ad = abs(d_hpf)
            if ad > self.pd_max:
                self.pd_max = ad
            self.tc_num += d_hpf * d_hpf
            self.tc_den += self.vel * self.vel
            self.win_n += 1
            if self.win_n >= WINDOW_SAMPLES:
                decision = self._close_window()

        # 6. State machine + window control (uses RAW cf for anti-tap gate)
        self._step_state(ratio, cf_raw, t_ms)

        rec = SampleRecord(
            timestamp_ms = (t_ms - self.start_ms) & 0xFFFFFFFF,
            sta_lta_ratio = ratio,
            cf_z          = cf,
            mpd_raw       = self.pd_max if self.window_open else 0.0,
            sample_count  = self.sample_count & 0xFFFF,
        )
        self.sample_count += 1
        return rec, decision

    # ─── Helpers ─────────────────────────────────────────────────────────
    def _open_window(self) -> None:
        self.win_n  = 0
        self.pd_max = 0.0
        self.tc_num = 0.0
        self.tc_den = 0.0
        self.window_open = True
        # Reset integrators per event (matches S3 firmware behaviour)
        self.hpf_a_in = self.hpf_a_out = 0.0
        self.vel       = 0.0
        self.hpf_v_in = self.hpf_v_out = 0.0
        self.disp      = 0.0
        self.hpf_d_in = self.hpf_d_out = 0.0

    def _close_window(self) -> Decision:
        pd = self.pd_max
        tc = (2.0 * math.pi * math.sqrt(self.tc_num / self.tc_den)
              if self.tc_den > 0 else 0.0)
        mw = self._estimate_mw(pd)
        self.window_open = False
        if mw >= self.mw_t2:
            return Decision(True, True, mw, pd, tc, "Mw_ge_T2")
        if mw >= self.mw_t1:
            return Decision(True, False, mw, pd, tc, "Mw_ge_T1")
        return Decision(False, False, mw, pd, tc, "below_threshold")

    def _estimate_mw(self, pd_m: float) -> float:
        if pd_m <= 0.0:
            return -99.0
        return MW_PD_A * math.log10(pd_m) + MW_PD_B * math.log10(self.r_km) + MW_PD_C

    def _step_state(self, ratio: float, cf_raw: float, t_ms: int) -> None:
        if self.state == State.STANDBY:
            if ratio >= RATIO_TRIGGER and cf_raw >= MIN_TRIG_CF_ABS:
                self.trig_count += 1
                if self.trig_count >= MIN_TRIG_COUNT:
                    self.state = State.DETECTING
                    self.peak_ratio = ratio
                    self._open_window()
            else:
                self.trig_count = 0

        elif self.state == State.DETECTING:
            if ratio > self.peak_ratio:
                self.peak_ratio = ratio
            if ratio < RATIO_DETRIGGER:
                self.state = State.STANDBY
                self.trig_count = 0
                self.peak_ratio = 0.0

        elif self.state == State.ALARMING:
            if t_ms - self.alarm_start_ms > ALARM_MAX_MS:
                self.state = State.LOCKOUT
                self.lockout_start_ms = t_ms

        elif self.state == State.LOCKOUT:
            if t_ms - self.lockout_start_ms > LOCKOUT_MAX_MS:
                self.state = State.STANDBY
                self.lta = self.lta_quiet
                self.sta = self.lta_quiet

    def seed_lta(self, noise_floor_cf: float) -> None:
        """Seed STA/LTA from an externally-measured noise floor (m²/s⁴)."""
        v = max(noise_floor_cf, LTA_FLOOR)
        self.sta = v
        self.lta = v
        self.lta_quiet = v

    # External hooks the orchestrator can call
    def force_alarm(self, t_ms: int) -> None:
        self.state = State.ALARMING
        self.alarm_start_ms = t_ms

    def reset_alarm(self, t_ms: int) -> None:
        self.state = State.STANDBY
        self.trig_count = 0
        self.peak_ratio = 0.0


# ─── CLI entrypoint: stdin CSV → stdout JSONL ────────────────────────────
def run_cli(in_stream: TextIO = sys.stdin, out_stream: TextIO = sys.stdout) -> int:
    algo = SeismoGuardAlgo()
    t = 0
    for raw in in_stream:
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        parts = line.split(",")
        if len(parts) < 3:
            continue
        try:
            ax = float(parts[0]); ay = float(parts[1]); az = float(parts[2])
        except ValueError:
            continue
        rec, decision = algo.process(ax, ay, az, t)
        out_stream.write(rec.to_json() + "\n")
        if decision is not None:
            out_stream.write(
                json.dumps({"decision": decision.reason,
                            "mw": round(decision.mw_est, 3),
                            "pd_m": decision.pd_m,
                            "tau_c": decision.tau_c}) + "\n")
        t += int(DT * 1000)
    return 0


if __name__ == "__main__":
    sys.exit(run_cli())
